flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java

```diff
@@ -695,6 +697,27 @@ protected void afterInvoke() throws Exception {
         // close all operators in a chain effect way
         operatorChain.closeOperators(actionExecutor);
 
+        // If checkpoints are enabled, waits for all the records get processed by the downstream
+        // tasks. During this process, this task could coordinate with its downstream tasks to
+        // continue perform checkpoints.
+        if (configuration.isCheckpointingEnabled()) {
+            LOG.debug("Waiting for all the records processed by the downstream tasks.");
+            CompletableFuture<Void> combineFuture =
+                    FutureUtils.waitForAll(
+                            Arrays.stream(getEnvironment().getAllWriters())
+                                    .map(
+                                            FunctionUtils.uncheckedFunction(
+                                                    ResultPartitionWriter
+                                                            ::getAllRecordsProcessedFuture))
+                                    .collect(Collectors.toList()));
+
+            MailboxExecutor mailboxExecutor =
+                    mailboxProcessor.getMailboxExecutor(TaskMailbox.MIN_PRIORITY);
+            while (!combineFuture.isDone()) {
```

Comment:
I would +1 for have a method to do blocking waiting should be more preferred. Considering the changes to the mailbox executor seems to be non-trivial, perhaps we could implement the mechanism separately, and for now we may first waiting with `future.get(1s)` like we do in [`StreamOperatorWrapper`](https://github.com/apache/flink/blob/84e1186d24ebbe6b3a4629496274d6337b333af1/flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamOperatorWrapper.java#L161) ?

Reply:
I don't think it works that way. The mailbox might be empty at the time you call `tryYield` (honestly I think in most cases it will) and you will immediately go for closing the task. If I understand correctly, at this point the mailbox will mostly process checkpoint triggers from time to time. Therefore we cannot close the executor immediately, however busy looping waiting for the `triggerCheckpoint` mail is also not the best option as they might be triggered occasionally.

Reply:
I think it would not go for closing the task directly if the mailbox is empty~? For each iteration, `while (mailboxExecutor.tryYield()) {}` would exit if no mails, then `future.get(1s)` would do a blocking waiting for 1 second, if the future still not finish, it would throws `TimeoutException` and get back to the outer iteration, Thus it seems to be a temporary fix similar to `sleep(1000)` that decreases the ratio of running on CPU~

Reply:
Ah right, I forgot/missed the outer loop. Yes, that would be one option. I don't like the sleep here, but I probably a better solution than busy looping. Not sure about the value of the sleep/wait. Maybe we could correlate it with the checkpoint interval? Still would like to hear from @pnowojski what he thinks about it.

Reply:
I've only very briefly took a look at this PR and this problem so far (I will take a one more look later), so take what I will write with a grain of salt :)

I think this is a bit hacky. It seems to me that in a proper solution, this should/could be handled by the regular default action (`StreamTask#processInput`) and spinning the mailbox via `runMailboxLoop()`. I think very similar problem was solved by Anton when he split the `runMailboxLoop()` between `StreamTask#executeRestore` and  `StreamTask#executeInvoke`. After the first one, mailbox is suspended and then restarted in the second one. It seems like you could have replaced this semi-busy loop with just another step (after `executeInvoke` mailbox should be suspended, and then restarted in the `afterInvoke()`)?

For reference you might want to take a look at the `MailboxProcessor#suspend`?

Reply:
I think that's actually a nice idea. We would need to change the `processInput` method slightly so that it suspends the `mailboxProcessor` instead of calling `allActionsComplete`. Then in the afterInvoke method we would call `allActionsComplete` once the `combineFuture` finishes.

```
    protected void processInput(MailboxDefaultAction.Controller controller) throws Exception {
        InputStatus status = inputProcessor.processInput();
        if (status == InputStatus.MORE_AVAILABLE && recordWriter.isAvailable()) {
            return;
        }
        if (status == InputStatus.END_OF_INPUT) {
            mailboxProcessor.suspend();
            return;
        }
        ....
    }

    protected void afterInvoke() throws Exception {
        LOG.debug("Finished task {}", getName());
        getCompletionFuture().exceptionally(unused -> null).join();

        final CompletableFuture<Void> timersFinishedFuture = new CompletableFuture<>();

        // close all operators in a chain effect way
        operatorChain.closeOperators(actionExecutor);

        // If checkpoints are enabled, waits for all the records get processed by the downstream
        // tasks. During this process, this task could coordinate with its downstream tasks to
        // continue perform checkpoints.
        final CompletableFuture<Void> taskCanFinish;
        if (configuration.isCheckpointingEnabled()) {
            LOG.debug("Waiting for all the records processed by the downstream tasks.");

            List<CompletableFuture<Void>> allRecordsProcessedFutures = new ArrayList<>();
            for (ResultPartitionWriter partitionWriter : getEnvironment().getAllWriters()) {
                partitionWriter.notifyEndOfUserRecords();
                allRecordsProcessedFutures.add(partitionWriter.getAllRecordsProcessedFuture());
            }
            taskCanFinish = FutureUtils.waitForAll(allRecordsProcessedFutures);
        } else {
            taskCanFinish = CompletableFuture.completedFuture(null);
        }
        taskCanFinish.thenRun(mailboxProcessor::allActionsCompleted);
        mailboxProcessor.runMailboxLoop();
 
        ....
    }

```

Reply:
Very thanks @pnowojski  @dawidwys  for the inputs! I also think we block on the mailbox loop would be nicer since it avoids the empty loop~ I'll update the PR with this method~